### PR TITLE
Add a list instance for the involutive monoid frexlet

### PR DESCRIPTION
**Pull request by yallop**
This is possibly a bit overdone: it would be significantly easier with propositional equality instead of setoids.

in any case, this PR adds list instances for monoids and involutive monoids along with tests (with explicit proof terms for the moment) for both instances.  For example,

```idris
test : {A : Setoid} -> {x,y : List (U A)} -> A .ListEquality (reverse (reverse y ++ ([] ++ reverse x)))
                                                             (x ++ y)
test = frexify (Frexlet.Monoid.Involutive.Frex.Frex List _) [x,y] ...
```

### CODE REVIEW

**Comment by ohad** [comment]
Thanks!

We should be able to reuse the stdlib proofs such as `Data.List.revOnto`.

We have this function for that purpose:
```idris
src/Data/Setoid/Definition.idr:reflect : (a : Setoid) -> {x, y : U a} -> x = y -> a.equivalence.relation x y
```

**Comment by yallop** [comment]
Ah, that's significantly easier.  Thanks!

**Comment by ohad** [comment]
Sorry about that --- we ought to write something like a paper documenting all of these.

**Comment by yallop** [comment]
8e2d74a adds a propositional version of the involutive instance (currently inlined within the test file), to check that proof search can construct the term if we dispense with setoids.  Happily, it can:

```idris
testProp : {A : Type} -> {x,y : List A} -> reverse (reverse y ++ ([] ++ reverse x)) = x ++ y
testProp = solve 2 (Frex PropListInvolutiveMonoid _)
         $ ((Dyn 1) .inv .*. (I1 .*. (Dyn 0) .inv)) .inv =-= Dyn 0 .*. Dyn 1
```

File: src/Frexlet/Monoid/List.idr [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import public Data.List
+import public Data.Setoid.List
+        
+%default total
+
+public export
+prodCong : {A:Setoid} -> (x1, x2, y1, y2 : List (U A)) ->
+  A .ListEquality x1 y1 -> A .ListEquality x2 y2 -> A .ListEquality (x1 ++ x2) (y1 ++ y2)
+prodCong [] x2 [] y2 Nil p = p
+prodCong (x::xs) x2 (y::ys) y2 (ph::pt) p = ph :: prodCong _ _ _ _ pt p
```
[the remark this answers is not in the document]
Comment (yallop) [reply src/Frexlet/Monoid/List.idr]: Also 59eb031

File: src/Frexlet/Monoid/Involutive/List.idr [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import Data.List
+
+%default total
+
+reverseHomomorphic : {A : Setoid} -> (x, y : List (U A)) -> (A .ListEquality x y) -> A .ListEquality (reverse x) (reverse y)
+reverseHomomorphic l1 l2 prf = roh [] [] l1 l2 (A .ListEqualityReflexive _) prf
+  where roh  : {A : Setoid} -> (w, x, y, z : List (U A)) -> (A .ListEquality w x) ->(A .ListEquality y z) ->
+               A .ListEquality (reverseOnto w y) (reverseOnto x z)
+        roh w x [] [] yz Nil = yz
+        roh w x (y::ys) (z::zs) wx (yzh :: yzt) = roh (y::w) (z::x) ys zs (yzh :: wx) yzt
```
[the remark this answers is not in the document]
Comment (yallop) [reply src/Frexlet/Monoid/Involutive/List.idr]: Good idea: 59eb031

**Pull request merged**